src/transfer_zip.py

```diff
@@ -0,0 +1,657 @@
+    if config.window is not None:
+        start_time = end_time - TimeDelta(config.window, format="quantity_str")
+    else:
+        start_time = Time(0, format="jd")
+    ok_timespan = Timespan(start_time, end_time)
+
+    # Find all exposures meeting criteria
+    dim_where = "(exposure.timespan OVERLAPS _ok_timespan)"
+    dim_where += f" AND ({data_query.where})" if data_query.where else ""
+    dim_bind = {"_ok_timespan": ok_timespan}
+    logger.info("Querying exposure: %s with %s", dim_where, dim_bind)
+    try:
+        exposures = source_butler.query_dimension_records(
+            "exposure",
+            where=dim_where,
+            bind=dim_bind,
+            limit=None,
+            instrument=data_query.instrument,
+        )
+    except EmptyQueryResultError:
+        logger.info("No matching records")
+        return
+    logger.info(
+        "Result is %d exposures from %s to %s",
+        len(exposures),
+        exposures[0].id,
+        exposures[-1].id,
+    )
+
+    for exp in exposures:
+        process_exposure(exp, data_query.instrument)
+
+
+def process_exposure(exp: DimensionRecord, instrument: str) -> None:
+    """Process an exposure by zipping, ingesting, and registering it in Rucio.
+
+    Parameters
+    ----------
+    exp: `lsst.daf.butler.DimensionRecord`
+        The exposure to process.
+    instrument: `str`
+        The name of the instrument corresponding to the exposure.
+    """
+    global logger, config, source_butler, dest_butler, rucio_interface
+
+    # Check several times for existence of the result to avoid work
+    # in case of race conditions
+    zip_name = f"{exp.obs_id}.zip"
+    dest_dir = (
+        ResourcePath(config.dest_uri_prefix).join(instrument).join(f"{exp.day_obs}")
+    )
+    dest_path = dest_dir.join(zip_name)
+    if dest_path.exists():
+        logger.info("Zip exists, skipping processing: %s", dest_path)
+        return
+
+    # Map exposure to tracts
+    tracts = set()
+    with source_butler.query() as q:
+        q = q.join_dimensions(["tract"]).where(
+            "visit = _exposure",
+            instrument=instrument,
+            skymap="lsst_cells_v1",
+            bind={"_exposure": exp.id},
+        )
+        for id in q.data_ids(["tract"]):
+            tracts.add(id["tract"])
+
+    # Find all datasets for this exposure and its source directory
+    refs = source_butler.query_datasets(
+        "raw",
+        exposure=exp.id,
+        instrument=instrument,
+        collections=f"{instrument}/raw/all",
+    )
+
+    logger.info("Handling exposure: %s, %s (%s)", exp.id, exp.obs_id, len(refs))
+
+    source_uri_dir = source_butler.getURI(refs[0]).dirname()
+    logger.debug("Source directory: %s", source_uri_dir)
+
+    # Make a zip file for this exposure
+    with tempfile.TemporaryDirectory() as tmpdir:
+        prepdir = os.path.join(tmpdir, "inputs")
+        os.mkdir(prepdir)
+        os.chdir(prepdir)
+        if dest_path.exists():
```

Comment:
Didn't you do this check a few lines earlier? Or is there a race condition that leads to lots of checks?

Reply:
The point is that I anticipate running many of these in parallel, possibly with overlapping time ranges, so I want to handle any race conditions properly.  That's also why I don't write directly to the raw destination but instead write locally and then copy.